File: SeeSpotHelp/StaceyAnimals/Animal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AnimalsNS
{
    public class Animal : Data.BaseObject
        
    {
 
        //Properties.
        public string id { get; set; }
        public string name { get; set; }
        public string type { get; set; }
        public string breed { get; set; }
        public string age { get; set; }
        public string groupId { get; set; }
        public StatusEnum status { get; set; }
        public string photo { get; set; }
        public string description { get; set; }

        public enum StatusEnum 
        {
            ADOPTABLE = 0,  // Animal is currently up for adoption.
            RESCUEONLY = 1,  // Animal can be adopted to rescue groups only.
            MEDICAL = 2,  // Animal is not up for adoption due to medical reasons.
            ADOPTED = 3,  // Animal has been adopted, YAY!
            PTS = 4,  // Animal has been put to sleep. :*(
            NLL = 5,  // Animal is No Longer Living due to other reasons.  Can be
            // used instead of PTS if people would prefer not to specify,
            // or if animal died of other causes.
            OTHER = 6 // In case I'm missing any other circumstances...
        }

    #region "Constructors"

        // Static constructor to initialize the static member
        static Animal() { }
        public Animal() { }

        public Animal(string aconnectionstring)
        {
            ConnectionString = aconnectionstring;
        }

        // Instance constructor that has the animal ID and is retrieved from Database.
        public Animal(string aconnectionstring, int ID)
        {
            ConnectionString = aconnectionstring;
           // getAnimal(ID);
        }

        public Animal(System.Data.DataRow row)
        {
            this.id = row["Id"].ToString();
            this.name = (string)row["name"];
            this.type = (string)row["type"];
            this.breed = (string)row["breed"];
            this.age = (string)row["age"];
            this.status = (StatusEnum)row["status"];
            this.groupId = (string)row["groupId"];
        }

        #endregion

        public static Animal InsertAnimal(string name,
                                          string type,
                                          string breed,
                                          string age,
                                          int status,
                                          string groupId)
        {
            object[] myParams = { name, "name",
                                  type, "type",
                                  breed, "breed",
                                  age, "age",
                                  status, "status",
                                  groupId, "groupId" };
            try
            {
                var animalData = Helpers.DBHelper.ExecuteProcedure(
                    Helpers.DBHelper.BuildConnectionString(), "Animal_Insert", myParams);
                if (animalData.Rows.Count > 0)
                {
                    return new Animal(animalData.Rows[0]);
                }
            }
            catch (Exception e)
            {
                System.Console.WriteLine(e.Message);
            }
            return null;
        }

        public static Animal UpdateAnimal(string animalId,
                                          string name,
                                          string type,
                                          string breed,
                                          string age,
                                          int status,
                                          string groupId)
        {
            object[] myParams = { animalId, "id",
                                  name, "name",
                                  type, "type",
                                  breed, "breed",
                                  age, "age",
                                  status, "status" };
            try
            {
                var animalData = Helpers.DBHelper.ExecuteProcedure(
                    Helpers.DBHelper.BuildConnectionString(), "Animal_Update", myParams);
                if (animalData.Rows.Count > 0)
                {
                    return new Animal(animalData.Rows[0]);
                }
            }
            catch (Exception e)
            {
                System.Console.WriteLine(e.Message);
            }
            return null;
        }

        //public void getAnimal(int myID)
        //{
        //    System.Data.DataTable dt;
        //    object[] myparams = { myID.ToString(), "animalid" };

        //    dt = Helpers.DBHelper.ExecuteProcedure(ConnectionString, "animal_get",myparams);

        //    if (dt.Rows.Count > 0)
        //    {
        //        initFromDR(dt.Rows[0]);
        //    } 

        //}

        //public void initFromDR(System.Data.DataRow dr)
        //{
        //    this.id = (int)dr["id"];
        //    this.name = (string)dr["name"];
        //    this.description = (string)dr["animaldescription"];

        //}
    }
}

File: SeeSpotHelp/SeeSpotHelp/App/core/animal.js
var DataServices = require('./dataservices');
var ServerResponse = require('./serverresponse');
var StringUtils = require('./stringutils');

// An animal that is currently being managed by a volunteer group.

var Animal = function(name, type, breed, age, status, photo, id, groupId) {
	this.name = name;
	this.type = type;
	this.breed = breed;
	this.age = age;
	this.status = status ? status : Animal.StatusEnum.ADOPTABLE;
	this.photo = photo ? photo : null;
	this.id = id ? id : null;
	this.groupId = groupId;


	// Unfortunately, I don't know anyway to generate this dynamically.
	this.classNameForSessionStorage = 'Animal';
};

Animal.GetTypeOptions = function () {
	return [
		"Dog", "Cat", "Other"
	];
};

Animal.prototype.CopyGroupFields = function (group) {
	// Add these fields for searching.
	this.zipCode = StringUtils.MakeSearchable(group.zipCode);
	this.shelter = StringUtils.MakeSearchable(group.shelter);
	this.city = StringUtils.MakeSearchable(group.city);
	this.state = StringUtils.MakeSearchable(group.state);
};

Animal.prototype.getPhoto = function() {
	if (this.photo) return this.photo;
	return this.type.toLowerCase() == "cat" ?
		"images/cat.png" :
		this.type.toLowerCase() == "dog" ?
		"images/dog.png" :
		"images/other.jpg";
};

Animal.StatusEnum = Object.freeze(
	{
		ADOPTABLE: 0,  // Animal is currently up for adoption.
		RESCUEONLY: 1,  // Animal can be adopted to rescue groups only.
		MEDICAL: 2,  // Animal is not up for adoption due to medical reasons.
		ADOPTED: 3,  // Animal has been adopted, YAY!
		PTS: 4,  // Animal has been put to sleep. :*(
		NLL: 5,  // Animal is No Longer Living due to other reasons.  Can be
		// used instead of PTS if people would prefer not to specify,
		// or if animal died of other causes.
		OTHER: 6 // In case I'm missing any other circumstances...
	}
);

Animal.castObject = function (obj) {
	var animal = new Animal();
	animal.copyFieldsFrom(obj);
	return animal;
};

Animal.prototype.copyFieldsFrom = function (other) {
	for (var prop in other) {
		this[prop] = other[prop];
	}
};

Animal.prototype.delete = function() {
	var firebasePath = "animals";
	DataServices.RemoveFirebaseData(firebasePath + "/" + this.id);
};

// Attempts to insert the current instance into the database as
// a animal
Animal.prototype.insert = function (callback) {
	var firebasePath = "animals";

	// Get rid of undefineds to make firebase happy.
	this.id = null;
	if (!this.photo) this.photo = null;

	this.id = DataServices.PushFirebaseData(firebasePath, this).id;
	DataServices.UpdateFirebaseData(firebasePath + "/" + this.id, { id: this.id });
};

Animal.prototype.update = function () {
	DataServices.UpdateFirebaseData("animals/" + this.id, this);
};

module.exports = Animal;
